### DISCUSSION TITLE
v1.5.x: Short-Term Roadmap

### DISCUSSION THREAD

**Discussion by JeodC** [DescentDevelopers/Descent3#94] [Ideas]
# v1.5.x: Short-Term Roadmap

So, the immediate goal for us maintainers was to get 1.5 back to a stable, modern release. That is to say, 1.5 was still in beta when development stalled and we view "1.5 Stable" as the equivalent to how the game might have been if that patch made it to retail. Now that the milestone is close, it's time to consider short-term roadmap options for "1.5.x", which we would like to focus on improvements. Here are several options we considered:

- Further modernize compiling and remove obsolete code
- Move to SDL2 (without relying on the icculus release)
- Integrate InjectD3 Quality of Life fixes like widescreen support, a more modern sound system, and input management
- Modernize netplay (remove IPX, add other protocols to get multiplayer functionality back up to snuff)*
- Integrate D3Edit into this primary repository (and add a compile-time option to build it alongside the game)
- Overhaul level scripts**

## Netplay modernization
I haven't attempted to play Descent 3's multiplayer in ages, so I'm not at all sure where the game stands when it comes to netcode and functionality. The above suggestion is my best guess. For example, TCPIP protocols likely need a second look and I assume UDP doesn't exist yet.

## Level scripts
Descent 3 levels can utilize scripts with C++ code, which make events like the training mission work properly. Without scripts, a level is mostly a sandboxed shell. Now C++ scripts wasn't a big concern in the 90s, but times are different. If scripts don't change, then the game will run unvetted, native C++ code when loading level scripts. Obviously, this allows potential for bad actors using Descent 3 levels as a medium.

Developers @kevinbentley and @midnite8177 have expressed interest in testing WasmEdge, a concept used by Microsoft Flight Simulator. WasmEdge executes code in a sandboxed environment, mitigating a risk. Other alternatives have also been mentioned in brief discussions: AngelScript, which the initial commit had bits and pieces of and is known for ease of integration into existing codebases, and Luascript, which is universal across platforms.

Each of these options has their strengths and weaknesses, and we're open to any of them. Our priorities lie in performance, ease of integration, and backwards compatibility.

**Comment by JeodC** [comment]
As usual we want the descent community at large to help guide the ship. Thanks for your participation and contributions!

**Comment by kevinbentley** [comment]
For the record, D3 uses only UDP, no TCP.

**Comment by JeodC** [comment]
Nice to know it wasn't *that* new

**Comment by MasterO2** [comment]
As far as modernizing netplay is concerned, is **IPv6 Support** also going to be added as well?

**Reply by JeodC** [reply]
Most likely, but that won't be for a long while.

**Reply by MasterO2** [reply]
> Most likely, but that won't be for a long while.

For your reference, both the Descent-Rebirth and Descent II-Rebirth source ports have IPv6 support:

https://github.com/dxx-rebirth/dxx-rebirth/ 

License is also **GPLv3**.

**Comment by Snyper587** [comment]
I'm a fan of being able to modify the FOV in Overload on my ultrawide at 3440x1440. Would be amazing if that could be implemented in D3. I know there's the aspect ratio adjustment command line switch, but that crops out some of the view.

**Reply by JeodC** [reply]
Test out @InsanityBringer's fork `Descent3Dev` which has some of those modernizations. Once it's polished up we'll look into it more--but 1.5.0 stable comes first. :)

**Reply by Snyper587** [reply]
Thanks @JeodC - Looks like I have to cancel my weekend plans. Been hoping for this for such a long time. Excited for what is to come for Descent 3!
![image0 (1)](https://github.com/DescentDevelopers/Descent3/assets/168311034/3e5c4f5f-f50d-4a7c-b7f1-d531a493b92b)
